plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -41,20 +45,66 @@ abstractly specifiable. It's an implementation detail.
 
 -}
 
+{- Note [Integer types for costing]
+We care about the speed of our integer operations for costing, this has a significant effect on speed.
+But we also need to care about overflow: the cost counters overflowing is a potential attack!
+
+We have a few choices here for what to do with an overflow:
+- Don't (this is what 'Integer' does, it's unbounded)
+- Wrap (this is what 'Int'/'Int64' and friends do)
+- Throw an overflow error (this is what 'Data.SafeInt' does)
+- Saturate (i.e. return max/min bound, this is what 'Data.SatInt does)
+
+In our case
+- Not overflowing would be nice, but 'Integer' is significantly slower than the other types.
+- Wrapping is quite dangerous, as it could lead to us getting attacked by someone wrapping
+their cost around to something that looks below the budget.
+- Throwing would be okay, but we'd have to worry about exception catching.
+- Saturating is actually quite nice: we care about whether `a op b < budget`. So long as `budget < maxBound`,
+  then `a op b < budget` will have the same truth value *regardless* of whether the operation overflows and saturates,
+  since saturating implies `a op b >= maxBound > budget`. Plus, it means we don't need to deal with
+  exceptions.
+
+So we use 'Data.SatInt', a variant of 'Data.SafeInt' that does saturating arithmetic.
+
+'SatInt' is quite fast, but not quite as fast as using 'Int64' directly (I don't know why that would be, apart from maybe
+just the overflow checks), but the wrapping behaviour of 'Int64' is unacceptable..
+
+One other wrinkle is that 'SatInt' is backed by an 'Int' (i.e. a machine integer with platform-dependent
+size), rather than an 'Int64' since the primops that we need are only available for 'Int' until GHC 9.2
+or so.
+
+However we mostly care about 64bit platforms, so this isn't too much of a problem. The only one where it could be a problem
+is GHCJS, which does present as a 32bit platform. However, we won't care about *performance* on GHCJS, since
+nobody will be running a node compiled to JS (and if they do, they deserve terrible performance). So:
+if we are not on a 64bit platform, then we can just fallback to the slower (but safe) 'Integer'.
+
+-}
+
+-- See Note [Integer types for costing]
+-- This relies on the fact that TH is run on the *target* platform, so even if we're e.g. cross-compiling to GHCJS
+-- this will give the right answer.
+type CostingInteger = $(if finiteBitSize (0::SatInt) < 64 then [t|Integer|] else [t|SatInt|])
```

Comment:
We could use CPP instead of TemplateHaskell:

```
{-# LANGUAGE CPP #-}
...
#include "MachDeps.h"

type CostingInteger = 
#if WORD_SIZE_IN_BITS < 64
    Integer
#else
   SatInt
#endif
```

Also I think  there is emulation of Int64 "primitives" for 32-bit systems using this: https://hackage.haskell.org/package/ghc-prim-0.7.0/docs/GHC-IntWord64.html
, but I haven't tried it.

Reply:
Yeah, but CPP is worse than TH, I think!

Reply:
> Also I think there is emulation of Int64 "primitives" for 32-bit systems using this

Not the ones we need, unfortunately.

Reply:
Actually I take it back, the TH is making HLS more unhappy than the CPP, so CPP it is :laughing: